### ISSUE TITLE
Energy prediction

### ISSUE THREAD

**Issue by Riq76** [Riq76/weas#1]

**Comment by Riq76** [comment]
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import MinMaxScaler
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, LSTM
import matplotlib.pyplot as plt
from datetime import datetime, timedelta

# Cargar datos desde un archivo Excel (se debe proporcionar el archivo "consumo_supermercado.xlsx")
data = pd.read_excel("consumo_supermercado.xlsx")

# Asegurarse de que los datos tienen las columnas necesarias
# Supongamos que las columnas son 'Fecha', 'Hora', 'Clima', 'Iluminacion', 'Frio_Alimentarios', 'Otros', 'Total_Tienda', 'Control_Clima', 'Control_Iluminacion'
data['FechaHora'] = pd.to_datetime(data['Fecha'] + ' ' + data['Hora'])
data = data.set_index('FechaHora')
data = data.resample('H').sum()  # Resampleo por hora si hay datos duplicados

# Normalizar los datos
scaler = MinMaxScaler()
columnas = ['Clima', 'Iluminacion', 'Frio_Alimentarios', 'Otros', 'Total_Tienda', 'Control_Clima', 'Control_Iluminacion']
data[columnas] = scaler.fit_transform(data[columnas])

# Crear secuencias de datos para entrenar el modelo
secuencia = 24  # Usar las últimas 24 horas para predecir la siguiente
X = []
y = []
for i in range(secuencia, len(data)):
    X.append(data[columnas].values[i-secuencia:i])
    y.append(data['Total_Tienda'].values[i])
X, y = np.array(X), np.array(y)

# Dividir los datos en conjunto de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Cambiar la forma de los datos para LSTM (samples, timesteps, features)
X_train = X_train.reshape((X_train.shape[0], X_train.shape[1], len(columnas)))
X_test = X_test.reshape((X_test.shape[0], X_test.shape[1], len(columnas)))

# Crear el modelo LSTM
modelo = Sequential()
modelo.add(LSTM(50, activation='relu', input_shape=(secuencia, len(columnas))))
modelo.add(Dense(1))
modelo.compile(optimizer='adam', loss='mse')

# Entrenar el modelo
historial = modelo.fit(X_train, y_train, epochs=20, batch_size=32, validation_data=(X_test, y_test), verbose=1)

# Evaluar el modelo
y_pred = modelo.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
print(f"Error cuadrático medio (MSE): {mse:.4f}")

# Desnormalizar los datos para la interpretación
y_test_desnormalizado = scaler.inverse_transform(y_test.reshape(-1, 1))
y_pred_desnormalizado = scaler.inverse_transform(y_pred)

# Visualizar resultados
plt.figure(figsize=(10, 6))
plt.plot(y_test_desnormalizado, label="Valores reales", color="blue")
plt.plot(y_pred_desnormalizado, label="Predicciones", color="red")
plt.title("Predicción de consumo de energía por hora usando LSTM con categorías y controles")
plt.xlabel("Índice de muestra")
plt.ylabel("Consumo de energía (kWh)")
plt.legend()
plt.show()

# Predicción para curvas con y sin señales de control
def predecir_curvas(data, modelo, scaler, columnas, secuencia, control_activo):
    consumo_futuro = []
    entrada = data[columnas].values[-secuencia:]

    for _ in range(8760 - len(data)):
        if not control_activo:
            entrada[:, -2:] = 0  # Desactivar las señales de control (Clima e Iluminación)
        entrada_reshaped = entrada.reshape((1, secuencia, len(columnas)))
        prediccion = modelo.predict(entrada_reshaped)[0, 0]
        consumo_futuro.append(prediccion)
        nueva_fila = entrada[-1].copy()
        nueva_fila[-2:] = 0 if not control_activo else nueva_fila[-2:]  # Mantener consistencia de controles
        nueva_fila[-3] = prediccion  # Usar la predicción como "Total_Tienda"
        entrada = np.append(entrada[1:], [nueva_fila], axis=0)

    consumo_futuro_desnormalizado = scaler.inverse_transform(np.array(consumo_futuro).reshape(-1, 1))
    return consumo_futuro_desnormalizado

# Predicciones con señales activas y desactivadas
curva_con_control = predecir_curvas(data, modelo, scaler, columnas, secuencia, control_activo=True)
curva_sin_control = predecir_curvas(data, modelo, scaler, columnas, secuencia, control_activo=False)

# Crear DataFrame con ambas predicciones
fechas_futuras = pd.date_range(data.index[-1] + timedelta(hours=1), periods=len(curva_con_control), freq='H')
resultado_df = pd.DataFrame({
    "FechaHora": fechas_futuras,
    "Consumo_Con_Control": curva_con_control.flatten(),
    "Consumo_Sin_Control": curva_sin_control.flatten()
})

# Guardar las predicciones en un archivo Excel
resultado_df.to_excel("prediccion_curvas_control.xlsx", index=False)
print("Predicciones futuras guardadas en 'prediccion_curvas_control.xlsx'")
